parlai/agents/starspace/modules.py

```diff
@@ -0,0 +1,84 @@
+# Copyright (c) 2017-present, Facebook, Inc.
+# All rights reserved.
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree. An additional grant
+# of patent rights can be found in the PATENTS file in the same directory.
+
+import math
+import torch
+import torch.nn as nn
+from torch.nn.parameter import Parameter
+from torch.autograd import Variable,Function
+from torch.nn.utils.rnn import pad_packed_sequence, pack_padded_sequence
+import torch.nn.functional as F
+
+class Starspace(nn.Module):
+    def __init__(self, opt, num_features, dict):
+        super().__init__()
+        self.lt = nn.Embedding(num_features, opt['embeddingsize'], 0,
+                               sparse=True, max_norm=opt['embeddingnorm'])
+        if not opt['tfidf']: 
+            dict = None
+        self.encoder = Encoder(self.lt, dict)
+        if not opt['share_embeddings']:
+            self.lt2 = nn.Embedding(num_features, opt['embeddingsize'], 0,
+                                    sparse=True, max_norm=opt['embeddingnorm'])
+            self.encoder2 = Encoder(self.lt2, dict)
+        else:
+            self.encoder2 = self.encoder
+        self.opt = opt
+
+    def forward(self, xs, ys=None, cands=None):
+        scores = None
+        if ys is not None:
+            # training
+            xs_enc = []
+            ys_enc = []
+            xs_emb = self.encoder(xs)
+            xs_enc.append(xs_emb)
+            ys_enc.append(self.encoder2(ys))
+            for c in cands:
+                xs_enc.append(xs_emb)
+                c_emb = self.encoder2(c)
+                ys_enc.append(c_emb)
+        else:
+            # testing
+            xs_enc = []
+            ys_enc = []
+            xs_emb = self.encoder(xs)
+            c_scores = []
+            for c in cands:
+                xs_enc.append(xs_emb)
+                c_emb = self.encoder2(c)
+                ys_enc.append(c_emb)
+        return torch.cat(xs_enc), torch.cat(ys_enc)
```

Comment:
maybe good to merge this since most of it is shared?

```
xs_enc = []
ys_enc = []
xs_emb = self.encoder(xs)
if ys is not None:
    xs_enc.append(xs_emb)
    ys_enc.append(self.encoder2(ys))
for c in cands:
    xs_enc.append(xs_emb)
    c_emb = self.encoder2(c)
    ys_enc.append(c_emb)
return torch.cat(xs_enc), torch.cat(ys_enc)
```

Reply:
ok